grudge/models/__init__.py

```diff
@@ -36,16 +38,26 @@ class Operator:
 class HyperbolicOperator(Operator):
     """A base class for hyperbolic Discontinuous Galerkin operators."""
 
-    def max_eigenvalue(self, t, fields, discr):
-        raise NotImplementedError
-
-    def estimate_rk4_timestep(self, discr, t=None, fields=None):
-        """Estimate the largest stable timestep for an RK4 method.
+    @abstractmethod
+    def max_characteristic_velocity(self, t, fields, dcoll):
+        """Return an upper bound on the characteristic
+        velocities of the operator.
         """
 
-        from grudge.dt_finding import (
-                dt_non_geometric_factor,
-                dt_geometric_factor)
-        return 1 / self.max_eigenvalue(t, fields, discr) \
-                * (dt_non_geometric_factor(discr)
-                * dt_geometric_factor(discr))
+    def estimate_rk4_timestep(self, dcoll, t=None, fields=None, dt_scaling=None):
+        """Estimate the largest stable timestep for an RK4 method."""
+        from mpi4py import MPI
+        from grudge.dt_utils import (dt_non_geometric_factor,
+                                     dt_geometric_factor)
+
+        max_lambda = self.max_characteristic_velocity(t, fields, dcoll)
+        dt_factor = \
+            (dt_non_geometric_factor(dcoll, scaling=dt_scaling)
+             * dt_geometric_factor(dcoll))
+
+        mpi_comm = dcoll.mpi_communicator
+        if mpi_comm is None:
+            return dt_factor * (1 / max_lambda)
+
+        return (1 / mpi_comm.allreduce(max_lambda, op=MPI.MAX)) \
+            * mpi_comm.allreduce(dt_factor, op=MPI.MIN)
```

Comment:
I think we'll also want a local-only version of this, i.e. something without the reduction baked in. `estimate_rk4_timestep_loc`?

x-ref: #113

Reply:
Marked to address this in a separate PR: https://github.com/inducer/grudge/pull/108/commits/f18a36970bd18dfe7e890d19813d97b9f0aed5ab